File: docs/standard/events/snippets/index/csharp/EventExample.cs
namespace EventsOverview;

// <DefineEvent>
class Counter
{
    public event EventHandler? ThresholdReached;

    protected virtual void OnThresholdReached(EventArgs e)
    {
        ThresholdReached?.Invoke(this, e);
    }

    // Provide remaining implementation for the class...
}
// </DefineEvent>

// <EventDataClass>
public class ThresholdReachedEventArgs : EventArgs
{
    public int Threshold { get; set; }
    public DateTime TimeReached { get; set; }
}
// </EventDataClass>

// <CustomDelegate>
public delegate void ThresholdReachedEventHandler(object sender, ThresholdReachedEventArgs e);
// </CustomDelegate>

File: docs/standard/events/snippets/index/vb/EventExample.vb
' <DefineEvent>
Public Class Counter
    Public Event ThresholdReached As EventHandler

    Protected Overridable Sub OnThresholdReached(e As EventArgs)
        RaiseEvent ThresholdReached(Me, e)
    End Sub

    ' Provide remaining implementation for the class...
End Class
' </DefineEvent>

' <EventDataClass>
Public Class ThresholdReachedEventArgs
    Inherits EventArgs

    Public Property Threshold As Integer
    Public Property TimeReached As Date
End Class
' </EventDataClass>

' <CustomDelegate>
Public Delegate Sub ThresholdReachedEventHandler(sender As Object, e As ThresholdReachedEventArgs)
' </CustomDelegate>
